feature (3dtiles): add support for batch table

NikoSaul opened the pull request:
This PR contribute to the 3D tiles support. issue #185
 - This PR parse the optional batch table for each supported format (.pnts, .b3dm) and make the relationship between an object and its Batch Table. The table is stored in the object for later usage (e.g: use batch ID to colorize mesh for example). 
 - See [Batch Table](https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/BatchTable/README.md) for more information

Example: 
![](http://zupimages.net/up/17/34/ctsm.png)
I move my mouse on 1 building and I have the information about him on your top left.

NikoSaul committed:
feature (3dtiles): add support for batch table
This commit parse the optional batch table for each supported format (.pnts, .b3dm). The table is stored in the object for later usage (e.g: use batch ID to colorize mesh for example)
See https://github.com/AnalyticalGraphicsInc/3d-tiles/blob/master/TileFormats/BatchTable/README.md for more information

NikoSaul committed:
feature: add in the gltfLoader the extension batch id
This commit add the ability to make the relationship between an object and its batchtable

peppsac commented:
Could you add a description in the PR text explaining what is the purpose of these changes?

Jeremy-Gaillard commented:
Could you modify the 3d-tiles example so that the batch table properties of the object you click on is displayed? It would make the purpose of the PR clearer.

NikoSaul committed:
test (3dtiles): add an example for watch the information about one building,
When you move your mouse on an object with batch table, you can display in the left of your screen the information about this object.

NikoSaul commented:
@peppsac I added an example to the 3dtiles.html.
You have just to move on a little building and you have the information on your left.

peppsac commented on examples/3dtiles.html:
```diff
@@ -80,7 +114,10 @@
             // -----------------------------------------------------
-            var $3dTilesLayerRequestVolume = new itowns.GeometryLayer('3d-tiles-request-volume', globe.scene);
+
+            const container = new itowns.THREE.Group();
```
You can't use `const` here (examples/* files are restricted to es5 for max compatibility)

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
 
             $3dTilesLayerRequestVolume.name = 'RequestVolume';
             $3dTilesLayerRequestVolume.url = 'https://raw.githubusercontent.com/AnalyticalGraphicsInc/3d-tiles-samples/master/tilesets/TilesetWithRequestVolume/tileset.json';
+            // $3dTilesLayerRequestVolume.url = 'http://localhost:8003/tilesets/TilesetWithRequestVolume/tileset.json';
```
You can remove this line.

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
+            </ul>
+        </div>
+
+        <script>
```
Why did you split the mouse move code handling to a separate `<script>` tag?

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
+                // Pick an object with batch id
+                const mouse = new itowns.THREE.Vector2((event.clientX / window.innerWidth) * 2 - 1, -(event.clientY / window.innerHeight) * 2 + 1);
+                const raycaster = new itowns.THREE.Raycaster();
+                const htmlInfo = document.getElementById('info');
```
You can't use `const` here (es5)

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
+            }
+            function picking() {
+                // Pick an object with batch id
+                const mouse = new itowns.THREE.Vector2((event.clientX / window.innerWidth) * 2 - 1, -(event.clientY / window.innerHeight) * 2 + 1);
```
You should pass the `event` var  to the `picking` function

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
+                    if (interAttributes) {
+                        if (interAttributes._BATCHID) {
+                            const face = intersects[i].face.a;
+                            const BatchID = interAttributes._BATCHID.array[face];
```
`batchID` (lowercase `b`)

peppsac commented on examples/3dtiles.html:
```diff
@@ -90,19 +127,87 @@
+
+                            // the html information
+                            Object.keys(batchInformation).map((objectKey) => {
+                                const info = batchInformation[objectKey] % 1 === 0 ? batchInformation[objectKey].toString() : batchInformation[objectKey].toFixed(3).toString();
```
Maybe add a comment explaining what is going on here...

peppsac commented on src/Core/Prefab/GlobeView.js:
```diff
@@ -332,8 +332,8 @@ GlobeView.prototype.removeLayer = function removeImageryLayer(layerId) {
 };
 
 GlobeView.prototype.selectNodeAt = function selectNodeAt(mouse) {
+    // update the picking ray with the camera and mouse position
```
Why did you modify this file?

peppsac commented on src/Renderer/ThreeExtended/B3dmLoader.js:
```diff
@@ -1,5 +1,6 @@
 import * as THREE from 'three';
 import './GLTFLoader';
+import BT from './BatchTable';
```
Another proposal:
```js
import './BatchTable';
```
And rename its method to avoid repetition:
```js
BatchTable.parse();
```
(assuming that `BatchTable` is always defined using json format)

nguyenign commented:
>You have just to move on a little building and you have the information on your left.

it's good but i prefer text pannel in scene like this example 
https://stemkoski.github.io/Three.js/Sprite-Text-Labels.html

peppsac reviewed (approved):
Looks good, thanks!

dudering commented:
First of all thanks a lot for creating iTowns! It's a really useful framework and I am glad I found it.

While browsing the examples I found the 3dtiles_loader.html example. Which works great with 3D Tilesets of buildings. The attributes of each building are shown on click.

However I did not figure out how to colorize the mesh of a clicked building based on the batch ID. 

Could you please let me know how this is possible?

jailln commented:
> First of all thanks a lot for creating iTowns! It's a really useful framework and I am glad I found it.
> 
> While browsing the examples I found the 3dtiles_loader.html example. Which works great with 3D Tilesets of buildings. The attributes of each building are shown on click.
> 
> However I did not figure out how to colorize the mesh of a clicked building based on the batch ID.
> 
> Could you please let me know how this is possible?

I @dudering , we recently replaced our own implementation by 3DTilesRendererJS for 3D tiles (see #2225). As you can see in the issue, one of the last remaining point is to reimplement 3D tiles stylization with this new layer (`OGC3DTilesLayer`) based on 3DTilesRendererJS. We still have it in mind and will do it soon. Meanwhile, if you want to easily colorize buildings you can use the old `C3DTilesLayer` and check the `widgets_3dtiles_style.html` example. (please note that this layer will disappear when we will have re-implemented the style for the new `OGC3DTilesLayer`)
